clang/lib/Interpreter/IncrementalParser.cpp

```diff
@@ -407,6 +406,16 @@ void IncrementalParser::CleanUpPTU(PartialTranslationUnit &PTU) {
       }
     }
   }
+
+  // FIXME: We should de-allocate MostRecentTU
+  for (Decl *D : MostRecentTU->decls()) {
+    if (!isa<NamedDecl>(D))
+      continue;
+    // Check if we need to clean up the IdResolver chain.
+    NamedDecl *ND = cast<NamedDecl>(D);
+    if (ND->getDeclName().getFETokenInfo())
```

Comment:
In which case wouldn't we have a frontend token here?

Reply:
If the `MostRecentTU` has a declaration which for some reason was removed from the IdResolver by exiting some lexical scope, for example? I suspect there are other language rules that may decide to make a declaration not visible to the C name lookup...

Reply:
Alright, so in this case `IdResolver` has no entry for `ND`? Can we assert that? Might also be a good a pointer for the uninformed reader.

Reply:
To clarify, you want an assert for the case where we try to remove the `ND` from the `IdResolved` but it was never registered? If so, `RemoveDecl` asserts (I was hurt couple of times there ;))